crates/transport/src/layers/cache.rs

```diff
@@ -0,0 +1,214 @@
+        let val = cache.get(key).cloned();
+        Ok(val)
+    }
+
+    /// Saves the cache to a file specified by the path.
+    /// If the files does not exist, it creates one.
+    /// If the file exists, it overwrites it.
+    pub fn save_cache(&self) -> TransportResult<()> {
+        let path = self.config.path.clone();
+        let file = std::fs::File::create(path).map_err(TransportErrorKind::custom)?;
+        let cache = self.cache.read().unwrap();
+
+        // Iterate over the cache and dump to the file.
+        let entries = cache
+            .iter()
+            .map(|(key, value)| FsCacheEntry { key: *key, value: value.clone() })
+            .collect::<Vec<_>>();
+        serde_json::to_writer(file, &entries).map_err(TransportErrorKind::custom)?;
+        Ok(())
+    }
+
+    /// Loads the cache from a file specified by the path.
+    /// If the file does not exist, it returns without error.
+    pub fn load_cache(&self) -> TransportResult<()> {
+        println!("Loading cache...");
+        let path = self.config.path.clone();
+        if !path.exists() {
+            println!("Cache file does not exist.");
+            return Ok(());
+        };
+        let file = std::fs::File::open(path).map_err(TransportErrorKind::custom)?;
+        let file = BufReader::new(file);
+        let entries: Vec<FsCacheEntry> =
+            serde_json::from_reader(file).map_err(TransportErrorKind::custom)?;
+        let mut cache = self.cache.write().unwrap();
+        for entry in entries {
+            cache.put(entry.key, entry.value);
+        }
+
+        println!("Loaded from Cache");
+        Ok(())
+    }
+}
+
+impl<S> Service<RequestPacket> for CachingService<S>
+where
+    S: Service<RequestPacket, Response = ResponsePacket, Error = TransportError>
+        + Send
+        + 'static
+        + Clone,
+    S::Future: Send + 'static,
+{
+    type Response = ResponsePacket;
+    type Error = TransportError;
+    type Future = TransportFut<'static>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), TransportError>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: RequestPacket) -> Self::Future {
+        let mut inner = self.inner.clone();
+        let this = self.clone();
+        match req.clone() {
+            RequestPacket::Single(ser_req) => {
+                let params_hash = ser_req.params_hash();
+                let resp = this.get(&params_hash);
+                match resp {
+                    Ok(Some(resp)) => {
+                        println!("Cache hit!");
+                        let raw = RawValue::from_string(resp).unwrap();
+                        let payload: ResponsePayload<Box<RawValue>, Box<RawValue>> =
+                            ResponsePayload::Success(raw);
+                        let response = Response { id: ser_req.id().clone(), payload };
+
+                        Box::pin(async move { Ok(ResponsePacket::Single(response)) })
```

Comment:
we can extract this to a new function that returns the rawvalue from the cache, if this is none, we send the request, if some we return